pkg/webreg/webreg.go

```diff
@@ -0,0 +1,455 @@
+	defer func() { logrus.Infof("rendered in %s", time.Now().Sub(start)) }()
+
+	w.Header().Set("Content-Type", "text/html;charset=UTF-8")
+
+	var mainPage = template.Must(template.New("mainPage").Funcs(
+		template.FuncMap{
+			"nameWithLink": func(name string) template.HTML {
+				return template.HTML(componentLink(name))
+			},
+			"linkedSteps": func(steps []string) template.HTML {
+				var builder strings.Builder
+				for i, step := range steps {
+					builder.WriteString(componentLink(step))
+					if i+1 < len(steps) {
+						builder.WriteString(" &rarr; ")
+					}
+				}
+				return template.HTML(builder.String())
+			},
+		},
+	).Parse(mainPageHTML))
+	refs, chains, wfs, docs := agent.GetRegistryComponents()
+	comps := generateComponentsStruct(refs, chains, wfs, docs)
+	fmt.Fprintf(w, htmlPageStart, "Step Registry Help Page")
+	if err := mainPage.Execute(w, comps); err != nil {
+		logrus.Errorf("Unable to render page: %v", err)
+	}
+	fmt.Fprintln(w, htmlPageEnd)
+}
+
+func MainPageHandler(agent load.RegistryAgent) http.HandlerFunc {
+	return func(w http.ResponseWriter, req *http.Request) {
+		trimmedPath := strings.TrimPrefix(req.URL.Path, req.URL.Host)
+		// remove leading slash
+		trimmedPath = strings.TrimPrefix(trimmedPath, "/")
+		// remove trailing slash
+		trimmedPath = strings.TrimSuffix(trimmedPath, "/")
+		if trimmedPath == "help" {
+			mainPageHandler(agent, w, req)
+			return
+		}
+		splitURI := strings.Split(trimmedPath, "/")
+		if len(splitURI) == 2 && splitURI[0] == "help" {
+			refs, chains, workflows, _ := agent.GetRegistryComponents()
+			if _, ok := refs[splitURI[1]]; ok {
+				referenceHandler(agent, w, req)
+				return
+			}
+			if _, ok := chains[splitURI[1]]; ok {
+				chainHandler(agent, w, req)
+				return
+			}
+			if _, ok := workflows[splitURI[1]]; ok {
+				workflowHandler(agent, w, req)
+				return
+			}
+		}
+		w.WriteHeader(http.StatusNotFound)
+		w.Write([]byte(http.StatusText(http.StatusNotFound)))
+	}
+}
+
+func syntaxBash(source string) (string, error) {
+	var output bytes.Buffer
+	lexer := lexers.Get("bash")
+	style := styles.Get("dracula")
+	formatter := html.New(html.Standalone(false))
+	iterator, err := lexer.Tokenise(nil, source)
+	if err != nil {
+		return "", fmt.Errorf("failed to tokenise source: %w", err)
+	}
+	err = formatter.Format(&output, style, iterator)
+	return output.String(), err
+}
+
+func referenceHandler(agent load.RegistryAgent, w http.ResponseWriter, req *http.Request) {
+	start := time.Now()
+	defer func() { logrus.Infof("rendered in %s", time.Now().Sub(start)) }()
+	w.Header().Set("Content-Type", "text/html;charset=UTF-8")
+	name := path.Base(req.URL.Path)
+
+	page := template.Must(template.New("referencePage").Funcs(
+		template.FuncMap{
+			"syntaxedSource": func(source string) template.HTML {
+				formatted, err := syntaxBash(source)
+				if err != nil {
+					logrus.Errorf("Failed to format source file: %v", err)
+				}
+				return template.HTML(formatted)
```

Comment:
We should not rely on `formatted` on error, should probably put some safe value there.